feat(fiscal): correctly calculate closing balance

jniles opened the pull request:
This commit adds a special route for the closing balance which no longer sums the entire period total before it.  Instead, it simply fetches period 0 of the next fiscal year.  I've also made some cosmetic changes by expanding the ui-grid to be larger and therefore more readable and using the bhFilterToggle component.

Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/3288

jniles committed:
feat(fiscal): correctly calculate closing balance

This commit adds a special route for the closing balance which no longer
sums the entire period total before it.  Instead, it simply fetches
period 0 of the next fiscal year.

mbayopanda reviewed (approved):
LGTM 👍 

The UI modification must also be applied to the fiscal opening balance page for having an harmonious UIs for both opening and closing balance page of a fiscal year.

jniles committed:
feat(balance): add option to include closing txn

This commit adds an option to the balance sheet to include the closing
transaction on the balance sheet.  Currently, the closing balance is
taken from period 0 of the forthcoming year, however, this does not
provide all the details necessary to work on the closing balance.

jniles committed:
chore(fiscal): create period 13 in procedure

This commit adds the ability to create period 13 in the Stored
Procedure.

jniles committed:
feat(fiscal): write closing txn to period 13

This commit writes the closing balance transaction to period 13 of the
current fiscal year for clarity.  If a fiscal year is closed, this
period will contain the balancing transaction.

Closes #3288.

jniles committed:
fix(test): update balance report tests

This commit updates the tests for the balance report after the period 13
changes made.  It also makes sure the fiscal year is closed correctly.

mbayopanda commented on server/models/procedures/time_period.sql:
```diff
@@ -76,13 +75,20 @@ BEGIN
       INSERT INTO period (`id`, `fiscal_year_id`, `number`, `start_date`, `end_date`, `locked`)
       VALUES (CONCAT(YEAR(fyStartDate), periodNumber), fiscalYearId, periodNumber, NULL, NULL, 0);
+
+    ELSEIF periodNumber = 13 THEN
```
does this line of code will work if the fiscal year has just 6 months ? 

the good one may be `fyNumberOfMonths + 1`

mbayopanda commented on server/models/procedures/time_period.sql:
```diff
@@ -141,6 +145,11 @@ BEGIN
 
+  -- get the current fiscal year date
+  SET currentFiscalYearPeriod13 = (
+    SELECT period.id FROM period WHERE period.fiscal_year_id = fiscalYearId AND period.number = 13 LIMIT 1
```
Could you use the max(period.number) of a fiscal year instead of 13

mbayopanda commented on server/controllers/finance/fiscal.js:
```diff
@@ -588,7 +637,7 @@ function getPeriodByFiscal(fiscalYearId) {
     FROM period
     WHERE period.fiscal_year_id = ?
-      AND period.number <> 13
+      AND period.number <> 0 AND period.number <> 13
```
Could you replace 13 by a variable name which refer the max(period.number) of the fiscal year, because in the case the fiscal year has just 6 months the period number will be 7 i think.

jniles replied:
That's a good idea.  I'll change all references to "Period 13" or "13" to "closing period".

mbayopanda commented on server/controllers/finance/reports/balance/index.js:
```diff
@@ -59,44 +62,70 @@ function document(req, res, next) {
 
+  // if we should include the closing balances, go all the way to period 13.
+  if (includeClosingBalances) {
+    const query = `${sql} WHERE fy.id = ? AND p.number = 13;`;
```
13 is a bad number 😆 

Could you use a variable which refer the max(period.number) of the fiscal year

mbayopanda commented on server/controllers/finance/reports/balance/report.handlebars:
```diff
@@ -16,11 +16,17 @@
+        <h4 class="text-center">
+          <strong class="text-capitalize">{{date period.fiscalYearStart "YYYY"}}</strong>
+        </h4>
+      {{else}}
```
👍

mbayopanda reviewed (commented):
The code LGTM but i made just some comment about the period 13

jniles committed:
chore(fiscal): rename period 13 to closing

This commit removes all references to period 13 and renames it "closing
period."   This renders the system flexible to have the fiscal years of
shorter length.
